### PR TITLE
[RLlib] Working/learning example: PPO + torch + LSTM.

### PR DESCRIPTION
**Pull request by sven1977**
RLlib torch currently does not generically support RNNs. Only QMIX (torch) has a built-in working LSTM. This PR provides a minimal working example for PPO+torch+RNN.

Addresses/Solves issue #7206 

Addresses/Solves issue #7206 

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: rllib/agents/ppo/ppo_torch_policy.py [hunk 1]
```diff
@@ -68,9 +68,10 @@ def __init__(self,
             use_gae (bool): If true, use the Generalized Advantage Estimator.
         """
         if valid_mask is not None:
+            num_valid = torch.sum(valid_mask)
 
             def reduce_mean_valid(t):
-                return torch.mean(t * valid_mask)
+                return torch.sum(t * valid_mask) / num_valid
```
Comment (ericl) [inline rllib/agents/ppo/ppo_torch_policy.py]: This is the same operation right? Or was it fixing a bug?

File: rllib/models/tf/__init__.py [hunk 2]
```diff
@@ -0,0 +1,14 @@
+# TODO(sven): Add once ModelV1 is deprecated and we no longer cause circular
+#  dependencies b/c of that.
```
Comment (ericl) [inline rllib/models/tf/__init__.py]: Hmm what's the advantage of making this a sub-package?

File: rllib/policy/eager_tf_policy.py [hunk 3]
```diff
@@ -313,10 +313,7 @@ def compute_actions(self,
             self._is_training = False
             self._state_in = state_batches
 
-            if tf.executing_eagerly():
-                n = len(obs_batch)
-            else:
-                n = obs_batch.shape[0]
+            n = len(obs_batch)
```
Comment (ericl) [inline rllib/policy/eager_tf_policy.py]: Doesn't len() not work in graph mode or am I missing something?

File: rllib/policy/rnn_sequencing.py [hunk 4]
```diff
@@ -14,13 +14,14 @@
...
-from ray.rllib.utils import try_import_tf
+from ray.rllib.utils import try_import_tf, try_import_torch
 
 tf = try_import_tf()
+torch, _ = try_import_torch()
 
 
 @DeveloperAPI
-def add_time_dimension(padded_inputs, seq_lens):
+def add_time_dimension(padded_inputs, seq_lens, framework="tf"):
```
Comment (ericl) [inline rllib/policy/rnn_sequencing.py]: Nice.

File: rllib/policy/torch_policy.py [hunk 5]
```diff
@@ -124,15 +141,63 @@ def compute_log_likelihoods(self,
...
+            parameters, _ = self.model(input_dict, state_batches, seq_lens)
             action_dist = self.dist_class(parameters, self.model)
             log_likelihoods = action_dist.logp(input_dict[SampleBatch.ACTIONS])
             return log_likelihoods
 
     @override(Policy)
     def learn_on_batch(self, postprocessed_batch):
-        train_batch = self._lazy_tensor_dict(postprocessed_batch)
+        shuffle = False
+        if shuffle:
```
Comment (ericl) [inline rllib/policy/torch_policy.py]: Isn't this always set to False in the line above?

Comment (ericl) [reply rllib/policy/torch_policy.py]: Maybe also comment this is copied from TF policy and vice versa (can we share the code in a util func)?

File: rllib/policy/torch_policy.py [hunk 6]
```diff
@@ -26,8 +27,15 @@ class TorchPolicy(Policy):
...
-                 action_distribution_class):
+    def __init__(self,
+                 observation_space,
+                 action_space,
+                 config,
+                 model,
+                 loss,
+                 action_distribution_class,
+                 max_seq_len=20,
+                 get_batch_divisibility_req=None):
```
Comment (ericl) [inline rllib/policy/torch_policy.py]: Hmm what's the reason it needs to be a lambda instead of a concrete value like in TF?

**Review by ericl — COMMENTED** [review]
Some questions for my own understanding.

File: rllib/agents/ppo/ppo_torch_policy.py (hunk 1 above)
Comment (sven1977) [reply rllib/agents/ppo/ppo_torch_policy.py]: No this was a bug. If you have `False`s in your mask, the mean will not be the mean of only the valid values, but the False-(=0)-diluted mean.

File: rllib/models/tf/__init__.py (hunk 2 above)
Comment (sven1977) [reply rllib/models/tf/__init__.py]: In general, everything should be a package, no? Looks nicer and allows for more intuitive imports.
Doing this always would enforce good class-hierarchy behavior. Currently, we have this circular dependency, which is why we cannot have this __init__.

File: rllib/policy/torch_policy.py (hunk 5 above)
Comment (sven1977) [reply rllib/policy/torch_policy.py]: Yeah, this `shuffle=False` is not properly done here. I'll move it into a util and try to share with tf.

File: rllib/policy/eager_tf_policy.py (hunk 3 above)
Comment (sven1977) [reply rllib/policy/eager_tf_policy.py]: I'm not sure. I just thought: Why would we check on tf.executing_eagerly() if in an eager policy? Are we not never in graph mode in that part of the code?

Comment (ericl) [reply rllib/policy/eager_tf_policy.py]: Oh hmm true. I think if tracing is on then sometimes you might end up in graph mode but not sure.

File: rllib/agents/ppo/ppo_torch_policy.py (hunk 1 above)
Comment (ericl) [reply rllib/agents/ppo/ppo_torch_policy.py]: Oof

File: rllib/models/tf/__init__.py (hunk 2 above)
Comment (ericl) [reply rllib/models/tf/__init__.py]: Fair enough.

File: rllib/policy/torch_policy.py (hunk 6 above)
Comment (sven1977) [reply rllib/policy/torch_policy.py]: Not sure. It's currently a lambda in tf and eager as well.

File: rllib/policy/torch_policy.py (hunk 5 above)
Comment (sven1977) [reply rllib/policy/torch_policy.py]: Done.

File: rllib/policy/eager_tf_policy.py (hunk 3 above)
Comment (sven1977) [reply rllib/policy/eager_tf_policy.py]: Yeah, it broke :) I'll leave it for now.

File: rllib/models/tf/__init__.py (hunk 2 above)
Comment (sven1977) [reply rllib/models/tf/__init__.py]: Cool.

File: rllib/agents/ppo/ppo_torch_policy.py (hunk 1 above)
Comment (sven1977) [reply rllib/agents/ppo/ppo_torch_policy.py]: It's ok. Torch LSTMs were never working anyways (hence this PR  😉) and the above code would only be executed in the RNN-case.

Comment (sven1977) [reply rllib/agents/ppo/ppo_torch_policy.py]: Torch doesn't have tf.boolean_mask, so I "implemented" it at first the wrong way with `t * valid_mask` when I did the torch-PPO translation. But again, only affected the RNN case, which wasn't function until now.

**Comment by sven1977** [comment]
Waiting for re-testing.
Example changed from Gru to LSTM (works literally 100 times better with LSTM :) )

**Comment by sven1977** [comment]
@ericl Please merge, tests are ok.

**Review by ericl — APPROVED** [review]
